docs/src/main/sphinx/develop/connectors.rst

```diff
@@ -297,6 +298,324 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
 error code and an actionable message, to let users know how to write a valid
 query.
 
+.. _dev-predicate-pushdown:
+
+Predicate pushdown
+^^^^^^^^^^^^^^^^^^
+
+When executing a query with a ``WHERE`` clause, the query plan can
+contain a ``ScanFilterProject`` operation with a predicate constraint.
+
+A predicate constraint is a description of the constraint imposed on the
+results of the query as expressed in the ``WHERE`` clause. For example,
+``WHERE x > 5 AND y = 3`` translates into a ``Constraint`` where the
+``summary`` field means the ``x`` column's domain must be greater than
+``5`` and the ``y`` column domain equals ``3``.
+
+When the query plan contains a ``ScanFilterProject`` operation, Trino
+tries to optimize the query by pushing down the predicate constraint
+into the connector by calling the ``applyFilter`` method of the
+connector metadata service. This method receives a table handle with
+all optimizations applied thus far, and should return either
+``Optional.empty()`` or a response with a new table handle derived from
+the old one.
+
+The query optimizer may call ``applyFilter`` for a single query multiple times,
+as it searches for an optimal query plan. It's important that connectors
+return ``Optional.empty()`` from ``applyFilter`` if they can't apply the
+constraint for this invocation even if they support ``ScanFilterProject``
+pushdown in general. It's also important to return ``Optional.empty()`` if the
+constraint has already been applied.
+
+A constraint contains:
+
+* a ``TupleDomain`` summary of possible values (or ranges) for different
+  columns,
+* an expression for pushing down function calls,
+* a map of assignments from variables in the expression to columns,
+* (optional) a predicate which tests a map of columns and their values,
+  it cannot be held on to after the ``applyFilter`` call returns
+* (optional) a set of columns the predicate depends on, must be present
+  if predicate is present.
+
+A ``TupleDomain`` defines a mapping between columns and their domains.
+A ``Domain`` is either a list of possible values or a list of ranges.
+It also contains information about nullability.
```

Comment:
A constraint contains the following elements:

* A ``TupleDomain`` defining the mapping between columns and their domains. A ``Domain`` is either a list of possible values, or a list of ranges, and also contains information about nullability.
* Expression for pushing down function calls.
* Map of assignments from variables in the expression to columns.
* (optional) Predicate which tests a map of columns and their values;
  it cannot be held on to after the ``applyFilter`` call returns.
* (optional) Set of columns the predicate depends on; must be present
  if predicate is present.